### DISCUSSION TITLE
[@types/react] Component built with @types/react@18 is not compatible with @types/react@17 when jsx=react-jsx is used

### DISCUSSION THREAD

**Discussion by kanoshin** [DefinitelyTyped/DefinitelyTyped#68968] [Issues with a @types package] [answered]
My company has a component library that aims to be compatible with both `react@17` and `react@18`. We use 18 as the version of react that gets installed for local development and in CI/CD. Here is the `package.json` file:
```
{
    "devDependencies": {
        "@types/react": "~18.2.55",
        "react": "~18.2.0",
        "typescript": "~4.7.4"
    },
    "peerDependencies": {
        "react": ">=17.0.2"
    },
}
```

We use `tsc` to compile TSX to JS with `"jsx": "react-jsx"` option. This is what `tsc` compiles as a typing for one of our components:
```ts
export declare const DefaultLoading: () => import("react/jsx-runtime").JSX.Element;
``` 

When trying to use such component in `react@17` codebase with `@types/react@17` we get compilation error:
```
Namespace '"node_modules/@types/react/jsx-runtime"' has no exported member 'JSX'.
```

This is because [jsx-runtime.d.ts](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/v17/jsx-runtime.d.ts) for react@17 file doesn't export standalone `JSX` namespace.

The best solution I could think of without dropping `"jsx": "react-jsx"` is to modify `jsx-runtime.d.ts` similarly to `react@18` typings:
```ts
// Expose `JSX` namespace in `global` namespace
import * as React from "./";
export { Fragment } from "./";

export namespace JSX {
    interface Element extends React.JSX.Element {}
    interface ElementClass extends React.JSX.ElementClass {}
    interface ElementAttributesProperty extends React.JSX.ElementAttributesProperty {}
    interface ElementChildrenAttribute extends React.JSX.ElementChildrenAttribute {}
    type LibraryManagedAttributes<C, P> = React.JSX.LibraryManagedAttributes<C, P>;
    interface IntrinsicAttributes extends React.JSX.IntrinsicAttributes {}
    interface IntrinsicClassAttributes<T> extends React.JSX.IntrinsicClassAttributes<T> {}
    interface IntrinsicElements extends React.JSX.IntrinsicElements {}
}

/**
 * Create a React element.
 *
 * You should not use this function directly. Use JSX and a transpiler instead.
 */
export function jsx(
    type: React.ElementType,
    props: unknown,
    key?: React.Key,
): React.ReactElement;

/**
 * Create a React element.
 *
 * You should not use this function directly. Use JSX and a transpiler instead.
 */
export function jsxs(
    type: React.ElementType,
    props: unknown,
    key?: React.Key,
): React.ReactElement;
```

My understanding is that such a change will not break existing v17 types and make components built with `@types/react@18` compatible with `@types/react@17`. Please note that with `"jsx": "react"` such components are already backward compatible, so the problem is isolated to `react-jsx` transform.

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
Thank you for the detailed report. This makes sense. Feel free to file a PR.

**Comment by kanoshin** [comment, marked as answer]
@eps1lon, thanks for approving the idea. Here is the PR: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68994
